src/bin/cargo-flamegraph.rs

```diff
@@ -174,28 +208,31 @@ fn build(opt: &Opt) {
         cmd.arg("--no-default-features");
     }
 
+    cmd.arg("--message-format=json-render-diagnostics");
+
     if opt.verbose {
         println!("build command: {:?}", cmd);
     }
 
-    let mut child = cmd
-        .spawn()
-        .expect("failed to spawn cargo build command");
+    let Output { status, stdout, .. } = cmd
+        .stderr(Stdio::inherit())
+        .output()
+        .expect("failed to execute cargo build command");
 
-    let exit_status = child.wait().expect(
-        "failed to wait for cargo build child to finish",
-    );
+    let artifacts: Vec<_> = Message::parse_stream(&*stdout)
+        .filter_map(|m| match m {
+            Ok(Message::CompilerArtifact(artifact)) => {
```

Comment:
This might be easier to folllow if you use `m.map()` and `collect()` into a `Result`, which you can then just expect on the outside -- maybe try it and see which one you think is easier to follow?

Reply:
I'd say it's a bit harder to read. But maybe that's just because I have to get used to it. While I like the `expect` on the outside, the `match` and `transpose` combo is a bit complex to read in my opinion. If you have any suggestions to simplify this, that would be great. The only way I see right now is to add an extension trait for `Message` that provides a `fn into_artifact(self) -> Option<Artifact>` but that seems like too much.

```rust
let artifacts = Message::parse_stream(&*stdout)
    .filter_map(|r| {
        r.map(|m| match m {
            Message::CompilerArtifact(a) => Some(a),
            _ => None,
        })
        .transpose()
    })
    .collect::<Result<Vec<_>, _>>()
    .expect("failed to parse cargo build output");
```

Reply:
That's fair -- go with the approach you like best.